### PR TITLE
Added methods from Jensen Group: graph-based genetic algorithm

### PR DESCRIPTION
**Pull request by sareeves96**

### CODE REVIEW

File: src/deriver/api.py [hunk 1]
```diff
@@ -13,6 +13,11 @@
 from .fragment_index import frag_index
 from .mate import mate
 import os
+import numpy as np
+from . import jensen_crossover as co
```
Comment (atomoton) [inline src/deriver/api.py:17]: variable names are too short and uninformative

File: src/deriver/api.py [hunk 2]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
+        if self.data.filter:
+            filter_params = self.data.filter_params
+        else:
+            logger.warning("Warning: No filter has been set, so all child molecules will be labeled"
+                           " as 'good' regardless of quality. Please call Deriver.set_filter() first"
+                           " in order to use a filter for drug-likeness.")
+            filter_params = None
+
+        for _ in range(n_children):
+            parent_a, parent_b = np.random.choice(self.data.seed_smiles)
```
Comment (atomoton) [inline src/deriver/api.py:378]: This only chooses one seed no?

File: src/deriver/api.py (hunk 1 above)
Comment (atomoton) [reply src/deriver/api.py:17]: also this import is not quite right... 

`from .jensen_crossover import method1, method2`

File: src/deriver/api.py [hunk 3]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
+            try:
+                new_child = co.crossover(parent_a, parent_b)
+                if new_child is not None:
+                    new_child = mu.mutate(new_child, mut_rate)
+                assert new_child
+                children.append(new_child)
+            except:
+                continue
```
Comment (atomoton) [inline src/deriver/api.py:386]: Way too broad (AssertionFailedError or something may work). Also at the very least needs to capture and log the exception before continuing.

File: src/deriver/api.py [hunk 4]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
+                if self.data.filter_molecules:
+                    if child not in self.data.filter_molecules:
+                        good_children.append(child)
+                    else:
+                        logger.debug(f"skipping previously seen molecule: {child}")
+                else:
+                    good_children.append(child)
+
+        logger.info(f"Generated {len(good_children)} 'good' children.")
+        return good_children, filtered_children
```
Comment (atomoton) [inline src/deriver/api.py:405]: needs to also add to a Deriver.Data field for molecules from this method, and to clear it at the beginning

File: src/deriver/api.py [hunk 5]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
```
Comment (atomoton) [inline src/deriver/api.py:447]: same .data comment

File: src/deriver/api.py (hunk 2 above)
Comment (atomoton) [reply src/deriver/api.py:378]: Okay yeah this should absolutely be random.sample

Comment (atomoton) [reply src/deriver/api.py:378]: `parent_a, parent_b = random.sample(self.data.seed_smiles, 2)`

File: src/deriver/api.py [hunk 6]
```diff
@@ -13,6 +13,11 @@
 from .fragment_index import frag_index
 from .mate import mate
 import os
+import numpy as np
```
Comment (atomoton) [inline src/deriver/api.py:16]: also, `numpy` random methods use a different seed than the `random` methods 

Try not to use them unless you need to, like if you are working with arrays or whatever

File: src/deriver/api.py [hunk 7]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
+        if self.data.filter:
+            filter_params = self.data.filter_params
+        else:
+            logger.warning("Warning: No filter has been set, so all child molecules will be labeled"
+                           " as 'good' regardless of quality. Please call Deriver.set_filter() first"
+                           " in order to use a filter for drug-likeness.")
+            filter_params = None
+
+        for _ in range(n_children):
+            parent_a, parent_b = np.random.choice(self.data.seed_smiles, size=2, replace=False)
```
Comment (atomoton) [inline src/deriver/api.py:420]:
```suggestion
            parent_a, parent_b = random.sample(self.data.seed_smiles, 2)
```

File: src/deriver/api.py (hunk 6 above)
Comment (atomoton) [inline src/deriver/api.py:16]:
```suggestion
import random
```

File: src/deriver/api.py [hunk 8]
```diff
@@ -357,6 +362,90 @@ def scan_selfies(self):
...
```
Comment (atomoton) [inline src/deriver/api.py:378]:
```suggestion
            parent_a, parent_b = random.sample(self.data.seed_smiles, 2)
```

File: src/deriver/api.py (hunk 3 above)
Comment (sareeves96) [reply src/deriver/api.py:392]: The exception handling I just added is the same as what is used in the original selfies deriver

**Pull request merged**